## 1. SETUP

In [1]:
from pathlib import Path
import json
import pandas as pd
import numpy as np

In [2]:
WORKING_DIR = Path.cwd()
DATA_DIR = WORKING_DIR / r"personal_style_postedit/logs"

print("Working Directory:", WORKING_DIR)
print("Dataset DIrectory: ", DATA_DIR)
print("Dataset exists:", DATA_DIR.exists())

Working Directory: d:\Academics Semester 5\DS501 - Machine Learning\Project\PersonalizationOfLLMs
Dataset DIrectory:  d:\Academics Semester 5\DS501 - Machine Learning\Project\PersonalizationOfLLMs\personal_style_postedit\logs
Dataset exists: True


In [3]:
json_files = sorted(DATA_DIR.glob("*.json"))

print("Number of JSON files:", len(json_files))
print("\nFirst 5 files:")
for f in json_files[:5]:
    print(f.name)

Number of JSON files: 81

First 5 files:
03acc614-1708-42a6-b142-9ef7a0f0ce4e.json
06285a67-fbc5-4673-9f1b-3e20aff3562f.json
0a303467-79b1-4641-9591-8c3a36f50413.json
0c529ac4-63ef-47e4-a8be-91a12f63242d.json
0d91417f-a097-424f-bf12-77b98654c26c.json


## 2. Inspect a JSON file

### 2.1 Inspect Keys of JSON dictionary

In [4]:
with open(json_files[0], "r", encoding = "utf-8") as f:
    participant = json.load(f)
    
print(type(participant))

if isinstance(participant, dict):
    print("\nTop level keys:")
    for key in participant.keys():
        print(key, end = ", ")

<class 'dict'>

Top level keys:
user_info, responses, pre_survey, mid_survey_1, mid_survey_2, post_survey, 

In [5]:
print("File: ", json_files[0].name)
print("Type: ", type(participant))

print("\nTop Level Keys: ")
for key in participant.keys():
    print(" -", key, end = ", ")
    print(type(key))

File:  03acc614-1708-42a6-b142-9ef7a0f0ce4e.json
Type:  <class 'dict'>

Top Level Keys: 
 - user_info, <class 'str'>
 - responses, <class 'str'>
 - pre_survey, <class 'str'>
 - mid_survey_1, <class 'str'>
 - mid_survey_2, <class 'str'>
 - post_survey, <class 'str'>


In [6]:
def inspect_structure(obj, indent=0, max_depth=3):
    prefix = " " * indent

    if indent // 2 >= max_depth:
        print(prefix + "...")
        return

    if isinstance(obj, dict):
        for key, value in obj.items():
            print(f"{prefix}{key}: {type(value).__name__}")
            inspect_structure(value, indent + 2, max_depth)

    elif isinstance(obj, list):
        print(f"{prefix}[list: {len(obj)} items]")
        if obj:
            inspect_structure(obj[0], indent + 2, max_depth)

    else:
        print(f"{prefix}{repr(obj)[:100]}")


inspect_structure(participant)

user_info: dict
  id: str
    '03acc614-1708-42a6-b142-9ef7a0f0ce4e'
  start_time: float
    1761748598.7927473
  conditions: list
    [list: 6 items]
      ...
  scenarios: list
    [list: 6 items]
      ...
responses: dict
  0: dict
    start_time: float
      ...
    scenario: str
      ...
    model_generation_shown: int
      ...
    details: str
      ...
    model_generation: str
      ...
    submit_details_time: float
      ...
    final_version: str
      ...
    edits: list
      ...
    submit_final_text_time: float
      ...
    likert: dict
      ...
  1: dict
    start_time: float
      ...
    scenario: str
      ...
    model_generation_shown: int
      ...
    details: str
      ...
    model_generation: str
      ...
    submit_details_time: float
      ...
    final_version: str
      ...
    edits: list
      ...
    submit_final_text_time: float
      ...
    likert: dict
      ...
  2: dict
    start_time: float
      ...
    scenario: str
      ...
    model_gen

### 2.2 JSON -> user_info

In [7]:
user_info = participant["user_info"]

for key, value in user_info.items():
    print(key, " : ", value)
    


id  :  03acc614-1708-42a6-b142-9ef7a0f0ce4e
start_time  :  1761748598.7927473
conditions  :  [1, 1, 1, 1, 0, 0]
scenarios  :  ['thank', 'wedding', 'apology', 'letter', 'condolence', 'reassurance']


### 2.3 JSON -> responses

In [8]:
response_0 = participant["responses"]["0"]

print("Scenario: ")
print(response_0["scenario"])

print("\nModel Generation Shown: ")
print(response_0["model_generation_shown"])

print("\nDetails: ")
print(response_0["details"])

print("\nModel Generation: ")
print(response_0["model_generation"])

print("\nFinal version: ")
print(response_0["final_version"])

print("\nNumber of raw edits:")
print(len(response_0["edits"]))

Scenario: 
thank

Model Generation Shown: 
1

Details: 
To my friend Julia
Emotional support
For always being there for me no matter what
I am beyond grateful for everything that you did do for me.
The best friend that i could ever asks for.

Model Generation: 
Dear Julia,

I’m writing to tell you how deeply grateful I am for your unwavering support over these past months. When life felt overwhelming, you were there to listen, to encourage, and to remind me of my strength. Your kindness, compassion, and endless patience have lifted my spirits more times than I can count. Whether I needed a shoulder to cry on or someone to celebrate the smallest victory, you never hesitated to step in.

You’ve shown me what true friendship looks like by standing by my side through every high and low. I value your laughter, your advice, and the countless moments you’ve brightened just by being present. I’m beyond grateful for every phone call, every hug, and every word of encouragement you’ve shared.

Th

In [9]:
print("All responses for participant:", participant["user_info"]["id"])

for task_id, response in participant["responses"].items():
    print(f"\n--- Response {task_id} ---")
    print("Scenario:", response["scenario"])
    print("Model generation shown:", response["model_generation_shown"])
    print("Model generation length:", len(response["model_generation"]))
    print("Final version length:", len(response["final_version"]))
    print("Number of edits:", len(response["edits"]))

All responses for participant: 03acc614-1708-42a6-b142-9ef7a0f0ce4e

--- Response 0 ---
Scenario: thank
Model generation shown: 1
Model generation length: 867
Final version length: 911
Number of edits: 96

--- Response 1 ---
Scenario: wedding
Model generation shown: 1
Model generation length: 835
Final version length: 842
Number of edits: 35

--- Response 2 ---
Scenario: apology
Model generation shown: 1
Model generation length: 810
Final version length: 925
Number of edits: 135

--- Response 3 ---
Scenario: letter
Model generation shown: 1
Model generation length: 845
Final version length: 859
Number of edits: 66

--- Response 4 ---
Scenario: condolence
Model generation shown: 0
Model generation length: 823
Final version length: 782
Number of edits: 936

--- Response 5 ---
Scenario: reassurance
Model generation shown: 0
Model generation length: 861
Final version length: 872
Number of edits: 1169


In [10]:
print("Participant:", participant["user_info"]["id"])

print("\nConditions:")
for i, condition in enumerate(participant["user_info"]["conditions"]):
    print(i, "->", condition)

print("\nScenarios:")
for i, scenario in enumerate(participant["user_info"]["scenarios"]):
    print(i, "->", scenario)

Participant: 03acc614-1708-42a6-b142-9ef7a0f0ce4e

Conditions:
0 -> 1
1 -> 1
2 -> 1
3 -> 1
4 -> 0
5 -> 0

Scenarios:
0 -> thank
1 -> wedding
2 -> apology
3 -> letter
4 -> condolence
5 -> reassurance


Each JSON file refers to one participant, who is asked to work on (work on means write OR edit) one text per scenario. 

There are 6 scenarios, so 6 texts of different themes. 

When condition is 1, it means the participant got an LLM generated text to edit. 
0 means he has to write his own text. 

This is useful, so that we could know - how does a person writes: what kind of sentence structures, the depth of vocabulary, sentence flow, word length, etc - could be learnt - and stored in a dense vector. 

Condition 1 can be used to understand how much is a person able to bring an AI generated text into his own writing. 

We have 81 participants. It means 81*6 = 486 Ai generated texts, and 486 Human edited or Human written texts. (Exact number unknown, until we find it using code)

#### Find Number of Human edited and Human Written Texts

In [11]:
import json
from pathlib import Path
from collections import Counter

#json_files should already contain 81 participant json paths
print("JSON files found: ", len(json_files)) # Though, already checked previously

participant_count = 0
response_count = 0

condition_counts = Counter()
shown_counts = Counter()
scenario_counts = Counter()

missing_fields = Counter()
bad_response_counts = []
scenario_mismatches = []
condition_shown_mismatches = []


JSON files found:  81


In [12]:
for path in json_files:
    
    # Open a file
    with open(path, "r", encoding = "utf-8") as f:
        data = json.load(f)
    
    # Participant Data - to record participants with != 6 responses
    participant_count += 1
    participant_id = data.get("user_info", []).get("id", path.stem)
    
    # Collect Conditions and Scenarios 
    # (used to count no. of conditions and scenarios)
    user_info = data.get("user_info", {})
    conditions = user_info.get("conditions", [])
    scenarios = user_info.get("scenarios", [])
    responses = data.get("responses", {})
    
    # Record Participants with != 6 responses
    if len(responses) != 6:
        bad_response_counts.append(
            (participant_id, len(responses))
        )
        
    for task_id, response in responses.items():
        
        response_count += 1
        i = int(task_id)
        
        # Condition and scenario assigned to this task
        condition = conditions[i] if i < len(conditions) else None
        assigned_scenario = scenario[i] if i < len(scenarios) else None
        
        # Information actually recorded in response
        actual_scenario = response.get("scenario")
        model_shown = response.get("model_generation_shown")
        
        condition_counts[condition] += 1
        shown_counts[model_shown] += 1
        scenario_counts[actual_scenario] += 1
        
        # Check Scenario alignment
        if assigned_scenario != actual_scenario:
            scenario_mismatches.append(
                (
                    participant_id,
                    task_id,
                    assigned_scenario,
                    actual_scenario
                )
            )
            
        # Check condition vs model_generation_shown
        if condition in (0, 1) and model_shown in (0, 1):
            if condition != model_shown:
                condition_shown_mismatches.append(
                    (
                        participant_id,
                        task_id,
                        condition,
                        model_shown
                    )
                )

        # Required fields
        for field in [
            "scenario",
            "model_generation",
            "final_version",
            "edits"
        ]:
            if field not in response or response[field] is None:
                missing_fields[field] += 1
        
        
        
    
    

In [13]:
print("\n========== DATASET SUMMARY ==========")

print("Participants:", participant_count)
print("Response/task records:", response_count)
print("Expected 6 × participants:", participant_count * 6)

print("\nCondition counts:")
for condition, count in sorted(condition_counts.items(), key=lambda x: str(x[0])):
    print(f"  Condition {condition}: {count}")

print("\nmodel_generation_shown counts:")
for value, count in sorted(shown_counts.items(), key=lambda x: str(x[0])):
    print(f"  {value}: {count}")

print("\nScenario counts:")
for scenario, count in sorted(scenario_counts.items(), key=lambda x: str(x[0])):
    print(f"  {scenario}: {count}")

print("\n========== VALIDATION ==========")

print("Participants without exactly 6 responses:",
      len(bad_response_counts))

print("Scenario mismatches:",
      len(scenario_mismatches))

print("Condition/model_shown mismatches:",
      len(condition_shown_mismatches))

print("\nMissing required fields:")
if missing_fields:
    for field, count in missing_fields.items():
        print(f"  {field}: {count}")
else:
    print("  None")

# --------------------------------------------------
# Calculate x
# --------------------------------------------------

x = condition_counts[1]
pure_human = condition_counts[0]

print("\n========== TEXT COUNTS ==========")

print("AI-generated texts stored:", response_count)
print("Human-edited texts (condition 1):", x)
print("Purely human-written texts (condition 0):", pure_human)
print("Total participant final texts:", response_count)

print("\nCheck:")
print(f"{x} + {pure_human} = {x + pure_human}")


========== DATASET SUMMARY ==========
Participants: 81
Response/task records: 487
Expected 6 × participants: 486

Condition counts:
  Condition 0: 162
  Condition 1: 324
  Condition None: 1

model_generation_shown counts:
  0: 162
  1: 324
  None: 1

Scenario counts:
  None: 1
  apology: 76
  condolence: 61
  eulogy: 19
  letter: 68
  reassurance: 60
  speech: 73
  thank: 78
  wedding: 51

========== VALIDATION ==========
Participants without exactly 6 responses: 1
Scenario mismatches: 486
Condition/model_shown mismatches: 0

Missing required fields:
  scenario: 1
  model_generation: 1
  final_version: 1
  edits: 1

========== TEXT COUNTS ==========
AI-generated texts stored: 487
Human-edited texts (condition 1): 324
Purely human-written texts (condition 0): 162
Total participant final texts: 487

Check:
324 + 162 = 486


Information:  Our scenario-alignment check was wrong

We got 486 Scenario mismatches. That doesn't mean all 486 records are corrupted.

It means our assumption that: user_info["scenarios"][i] corresponds to: responses[str(i)]["scenario"] is false.

We would have to find the scenario that response[str(i)] works on by deliberately reading responses[str(i)]["scenario"]

The response's own "scenario" field is the authoritative task identifier; we shouldn't compare it against the user_info["scenarios"] list positionally.

Good News: There is no error in which condition indicates 0, but model generationw as shown to the user for editing, or vice versa. So, we don't need to worry about it.

Problem: We have one participant with 7 responses. Hence, we have 487 responses instead of 486. Lets find that esteemed participant.

In [14]:
ans = None

for path in json_files:
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)

    responses = data.get("responses", {})
    
    if len(responses) != 6:
        print("FILE:", path)
        ans = path
        print("PARTICIPANT:", data.get("user_info", {}).get("id"))
        print("NUMBER OF RESPONSES:", len(responses))
        print("RESPONSE KEYS:", list(responses.keys()))
        
        for task_id, response in responses.items():
            print(
                task_id,
                "| scenario =", response.get("scenario"),
                "| condition =", (
                    data.get("user_info", {})
                        .get("conditions", [None] * len(responses))[int(task_id)]
                    if str(task_id).isdigit()
                    and int(task_id) < len(
                        data.get("user_info", {}).get("conditions", [])
                    )
                    else None
                ),
                "| shown =", response.get("model_generation_shown"),
                "| final exists =", bool(response.get("final_version")),
            )

FILE: d:\Academics Semester 5\DS501 - Machine Learning\Project\PersonalizationOfLLMs\personal_style_postedit\logs\417a64ff-e3b5-45c1-a926-83e8b3897840.json
PARTICIPANT: 417a64ff-e3b5-45c1-a926-83e8b3897840
NUMBER OF RESPONSES: 7
RESPONSE KEYS: ['0', '1', '2', '3', '4', '5', '7']
0 | scenario = apology | condition = 0 | shown = 0 | final exists = True
1 | scenario = wedding | condition = 0 | shown = 0 | final exists = True
2 | scenario = condolence | condition = 1 | shown = 1 | final exists = True
3 | scenario = letter | condition = 1 | shown = 1 | final exists = True
4 | scenario = speech | condition = 1 | shown = 1 | final exists = True
5 | scenario = thank | condition = 1 | shown = 1 | final exists = True
7 | scenario = None | condition = None | shown = None | final exists = False


In [15]:
with open(ans, "r", encoding = "utf-8") as f:
    data = json.load(f)

print(data.get("responses").get("7"))

{'start_time': 1761765499.0132513}


You will be surprised to know how much it is written .

{'start_time': 1761765499.0132513} 

Meaning, it was probably a bug, because of which another task was opened. Without any theme or aim.

But we will not delete or modify the original JSON file. We will just write a code which will ignore response records without a scenario.

The "7" record is not a seventh writing task. It is an empty/malformed response entry and should be excluded from the analysis.

So, our actual dataset consists of 486 AI generated outputs.
Then 486 Human outputs, out of which 324 are human edited outputs (Condition 1), and 162 are purely human written.

## 3. Create Train-Test Split

We split 81 participants into 57 participants for training dataset and 24 participants for testing dataset

### 3.1 Create A3 analysis dataframe

In [16]:
records = list()

for path in json_files:
    with open(path, "r", encoding = "utf-8") as f:
        data = json.load(f)
    
    participant_id = data.get("user_info", {}).get("id", {})
    responses = data.get("responses", {})
    
    
    for task_id, response in responses.items():
        if not response.get("scenario"):
            continue
        
        condition = response.get("model_generation_shown")
        
        records.append({
            "participant_id": participant_id,
            "task_id": task_id,
            "scenario": response["scenario"],
            "condition": condition,
            "ai_text": response.get("model_generation"),
            "final_text": response.get("final_version"),
            "edits": response.get("edits", [])
        })


In [17]:
df = pd.DataFrame(records)

print("Number of valid task records:", len(df))
print("Number of participants:", df["participant_id"].nunique())

print("\nCondition counts:")
print(df["condition"].value_counts().sort_index())

print("\nScenario counts:")
print(df["scenario"].value_counts().sort_index())

print("\nSum of such counts: ", df["scenario"].value_counts().sort_index().sum())


print("\nMissing values:")
print(df[["ai_text", "final_text"]].isna().sum())

df.head()

Number of valid task records: 486
Number of participants: 81

Condition counts:
condition
0    162
1    324
Name: count, dtype: int64

Scenario counts:
scenario
apology        76
condolence     61
eulogy         19
letter         68
reassurance    60
speech         73
thank          78
wedding        51
Name: count, dtype: int64

Sum of such counts:  486

Missing values:
ai_text       0
final_text    0
dtype: int64


,participant_id,task_id,scenario,condition,ai_text,final_text,edits
0,03acc614-1708-42a6-b142-9ef7a0f0ce4e,0,thank,1,"Dear Julia,\n\nI’m writing to tell you how dee...","Dear Julia,\n\nI’m writing to tell you how dee...","[{'type': 'delete', 'position': 111, 'removed'..."
1,03acc614-1708-42a6-b142-9ef7a0f0ce4e,1,wedding,1,"To MJ,\n\nI promise to cherish every laugh we ...","To MJ,\n\nI promise to cherish every laugh we ...","[{'type': 'delete', 'position': 443, 'removed'..."
2,03acc614-1708-42a6-b142-9ef7a0f0ce4e,2,apology,1,"Dear Bok,\n\nI want to apologize for taking yo...","Dear Bok,\n\nI want to apologize for taking yo...","[{'type': 'insert', 'position': 810, 'removed'..."
3,03acc614-1708-42a6-b142-9ef7a0f0ce4e,3,letter,1,"Dear Nom,\n\nI hope you’re doing well. I wante...","Dear Nom,\n\nI hope you’re doing well. I wante...","[{'type': 'insert', 'position': 845, 'removed'..."
4,03acc614-1708-42a6-b142-9ef7a0f0ce4e,4,condolence,0,"Dear J,\n\nMy deepest condolences for your los...","Dear J,\n\nMy deepest condolences to you and y...","[{'type': 'insert', 'position': 0, 'removed': ..."


### 3.2 - Inspect remaining data fields
(we akready saw user_info and edits)

**USER_INFO** -
 
1. id: Participant ID
2. start_time
3. conditions (Whether this scenario is purely human written or Ai text which is being edited)
4. scenarios

**RESPONSES** - 

Subdictionaries - "0", "1", "2", "3", "4", "5"

Subdictionary of each "0" and so on:
1. start_time
2. scenario
3. model_generation_shown
4. details
5. model_generation
6. submit_details_time
7. final_version
8. edits

Edits
1. type (insert, delete)
2. position
3. removed
4. added
5. timestamp

In [18]:
# Inspect the other text fields available for one participant

participant_id = df["participant_id"].iloc[0]

for path in json_files:
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)

    if data.get("user_info", {}).get("id") == participant_id:
        print("Participant:", participant_id)
        print("\nTop-level keys:")
        print(list(data.keys()))

        print("\nPre-survey:")
        print(data.get("pre_survey"))

        print("\nMid-survey 1:")
        print(data.get("mid_survey_1"))

        print("\nMid-survey 2:")
        print(data.get("mid_survey_2"))

        print("\nPost-survey:")
        print(data.get("post_survey"))

        break

Participant: 03acc614-1708-42a6-b142-9ef7a0f0ce4e

Top-level keys:
['user_info', 'responses', 'pre_survey', 'mid_survey_1', 'mid_survey_2', 'post_survey']

Pre-survey:
{'conf': '4', 'likely_i': '4', 'likely_ni': '3'}

Mid-survey 1:
{'mental': '6', 'hard': '13', 'insecure': '1', 'performance': '1', 'temporal': '18', 'condition': 1, 'submit_time': 1761751497.8881862}

Mid-survey 2:
{'mental': '21', 'hard': '21', 'insecure': '16', 'performance': '1', 'temporal': '21', 'condition': 0, 'submit_time': 1761753401.7445922}

Post-survey:
{'conf': '5', 'likely_i': '5', 'likely_ni': '1', 'future_pref': 'postedit', 'rankings': '["Originality","Style","Overall_quality","Efficiency","Reliability","Privacy","Ownership"]', 'gender': 'woman', 'trans': 'cis', 'hisp': 'No', 'race': 'Black or African American', 'age': 29, 'loe': 'College graduate (4 or 5 year program)', 'gender_table': '4', 'race_table': '1', 'age_table': '2', 'loe_table': '1', 'present_other': 'None', 'mask_other': 'None', 'feedback': 'I

**PRE SURVEY**
1. conf - Confidence in ability to edit AI weritten text to match his own style (0 - 5) 
2. likely_i - Likely to use Ai when capturing his voice is important (0 - 5)
3. likely_ni - Likely to use AI when capturing voice not important (0 - 5)

**MID_SURVEY 1 and MID_SURVEY 2**

1. mental - Mental Demand to write tasks (0-20)
2. hard - How hard did you have to work to write at your level of performance? (0-20)
3. insecure - Hpw annoyed/discourages/stressed/insecure felt (0-20)
4. performance - How successful were you in the writing tasks?" (0-20) (Very successful - Very unsuccessful)
5. temporal - how long they felt like the task was taking (0-20) (Very SHort - Very long)

CONDITION

0. means Human writing
1. means AI post editing

**POST SURVEY**

1. 'conf'
2. 'likely_i'
3. 'likely_ni'

4. 'future_pref': 'postedit', 'alone', 'model'

5. 'rankings': '["Originality","Style","Overall_quality","Efficiency","Reliability","Privacy","Ownership"]'
Rank the seven factors from most influential to least influential in determining their preference for how they would write in the future.

6. 'gender' - woman/man/non-binary
7. 'trans' - trans/cis
8. 'hisp' - yes/no: Hispanic? Or not?

9. 'race'   - American Indian or Alaska Native
            - Asian
            - Black or African American
            - Native Hawaiian or Other Pacific Islander
            - White


10. 'age'


11. 'loe'   - Didn’t Finish High School
            - Didn’t Finish High School but completed a technical/vocational program
            - High School Graduate or GED (General Education Diploma)
            - Completed High School and a technical/vocational program
            - Less than 2 years of college
            - 2 Years of College or more/ including associate degree or equivalent
            - College graduate (4 or 5 year program)
            - Master’s degree (or other post-graduate training)
            - Doctoral degree (PhD, MD, JD, etc.))

12. 'gender_table'
13. 'race_table'
14. 'age_table'
15. 'loe_table'

How important it is to the user that this part of their identity does or does not come across in the writing tasks completed in this study.

1 is "Very important that it does not come across"
3 is "It doesn't matter to me how readers perceive this characteristic"
5 is "Very important that it does come across"

16. 'present_other'
17. 'mask_other'

18. 'feedback'
19. 'submit_time'

#### An important discovery for A3

The reference paper saus tje study compares treatment text ahaomst the participant's control writing. 

In Anthropology paper, the section H1a talks about making AI-assisted writing more smilar to his pure human writing.

The methodology says that we calculate similarity by **cosine similarity between their LUAR embeddings**.

We can calculate **Similarity(post-edited final, pure participant's writing)**

In [19]:
participant_summary = (
    df.groupby(["participant_id", "condition"])
    .size()
    .unstack(fill_value = 0)
    .rename(columns = {0: "control_count", 1: "treatment_count"})
)

print(participant_summary.head(10))

print("\nControl-count distribution:")
print(participant_summary["control_count"].value_counts().sort_index())

print("\nTreatment-count distribution:")
print(participant_summary["treatment_count"].value_counts().sort_index())

condition                             control_count  treatment_count
participant_id                                                      
03acc614-1708-42a6-b142-9ef7a0f0ce4e              2                4
06285a67-fbc5-4673-9f1b-3e20aff3562f              2                4
0a303467-79b1-4641-9591-8c3a36f50413              2                4
0c529ac4-63ef-47e4-a8be-91a12f63242d              2                4
0d91417f-a097-424f-bf12-77b98654c26c              2                4
0e4ed97b-533a-4c35-9268-2610361186c3              2                4
1460c925-65dc-4f85-9bee-a3029020dd78              2                4
16e11b3b-7143-40ed-ba2c-c94e96d5be86              2                4
1a25aba9-d542-49a4-94df-900c081551b6              2                4
1a6bc3bd-868e-47cf-970d-67cf7d516465              2                4

Control-count distribution:
control_count
2    81
Name: count, dtype: int64

Treatment-count distribution:
treatment_count
4    81
Name: count, dtype: int64


## Phase 1 - Train Test Split

57 participants are assigned to the training/fit set and 24 participants are assigned to the held-out/test set.

In [20]:
from pathlib import Path

json_files = sorted(
    Path(DATA_DIR).rglob("*.json")
)

print("Total JSON Files: ", len(json_files))

Total JSON Files:  81


In [21]:
import random
random.seed(42)

shuffled_files = json_files.copy()
random.shuffle(shuffled_files)

train_files = shuffled_files[:57]
test_files = shuffled_files[57:]

print("Training files: ", len(train_files))
print("Testing files: ", len(test_files))

Training files:  57
Testing files:  24


In [22]:
train_set = set(train_files)
test_set = set(test_files)

print("Total Unique Files: ", len(train_set.union(test_set)))
print("Overlapping Files: ", len(train_set.intersection(test_set)))

Total Unique Files:  81
Overlapping Files:  0


In [23]:
train_file_list = [str(p) for p in train_files]
test_file_list = [str(p) for p in test_files]

with open("train_files.txt", "w", encoding = "utf-8") as f:
    f.write("\n".join(train_file_list))
    
with open("test_files.txt", "w", encoding = "utf-8") as f:
    f.write("\n".join(test_file_list))

Each participant has exactly 2 control counts (pure human written texts) and 4 treatment counts (human edited llm generated texts).

## Phase 2 - Construct Human Style Reference

Each participant has 2 control texts (pure human written) and 4 treatment texts (Ai generated human edited).

We intend to generate an embedding from the participant's control texts as well as his treatment texts.

And then we would find LUAR similarity between the reference control text and his individual treatment texts. This would give us how much is the person able to bring an AI generated text into his own writing style.

How do we generate an embedding? We concatenate the two control texts, and then find its embedding.

Participant_id

control_text_1

control_text_2

control_reference



For each treatment i (Treatment 1, Treatment 2, Treatment 3 and Treatment 4), we would find 
1. LUAR embedding of AI draft (Lets call it AI_i)
2. LUAR embedding of final draft (Lets call it final_i)
3. Find Similarity between AI_i and control embedding
4. Find similarity between final_i and control embedding
5. Subtract the latter from the former 

if the difference is positive, it means that the participant's post edited text moved closer to their own control-writing representation.

### How does it happen in code?

1. Raw tezt
2. LUAR tokenizer
3. LUAR-MUD model
4. 512 dimensional vector

We are using the pretrained/frozen LUAR-MUD model, not training LUAR on our 81 participants. The official model was trained on the Reddit Million User Dataset

In [24]:
from transformers import AutoTokenizer, AutoModel

MODEL_NAME = "rrivera1849/LUAR-MUD"

tokenizer = AutoTokenizer.from_pretrained(
    MODEL_NAME,
    trust_remote_code=True
)

model = AutoModel.from_pretrained(
    MODEL_NAME,
    trust_remote_code=True
)

model.eval()

print("LUAR loaded successfully!")

c:\Users\Samarth PC\AppData\Local\Programs\Python\Python313\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Could not reach the Hub ([Errno 11001] getaddrinfo failed). Using cached commit hash for 'rrivera1849/LUAR-MUD'.
Could not reach the Hub ([Errno 11001] getaddrinfo failed). Using cached commit hash for 'rrivera1849/LUAR-MUD'.
Loading weights: 100%|██████████| 105/105 [00:00<00:00, 1030.14it/s]

LUAR loaded successfully!


#### **METHODOLOGY**

I want to load LUAR model and generate embeddings - let us first do it for one participant. 

This includes :
1. 1 embedding of the AI text (how Ai writes - we concatenate 6 AI texts and find its embedding). 
2. Embeding of his pure human writing (concatenate 2 oure human writings and then find its embedding)
3. The 4 treatment embeddings - 1 each for the 4 treatment texts. 

Then I calculate and return the % to which the participant was able to get Ai generated Text A, B, C and D into his own writing. 

**This will be an entire function.**

### Phase 2.1 - Function to generate LUAR Embedding
Generate a single 512-dimensional LUAR-MUD embedding for one text document.


In [25]:
import torch


def luar_embedding(text, max_length=512):
    # Tokenize one document
    inputs = tokenizer(
        [text],
        return_tensors="pt",
        padding="max_length",
        truncation=True,
        max_length=max_length
    )

    # LUAR expects:
    # (batch_size, episode_length, sequence_length)
    inputs["input_ids"] = inputs["input_ids"].reshape(1, 1, -1)
    inputs["attention_mask"] = inputs["attention_mask"].reshape(1, 1, -1)

    # No gradients needed for inference
    with torch.inference_mode():
        embedding = model(**inputs)

    # Shape: (1, 512) → (512,)
    return embedding.squeeze(0).cpu()

In [26]:
with open(train_files[10], "r", encoding = "utf-8") as f:
    participant = json.load(f)
    
print(list(participant.keys()), "\n")

print(participant.get("user_info", {}).get("id"))
print(list(participant.get("responses", {}).keys()))
print(list(participant.get("responses", {}).get("0", {}).keys()), "\n\n")
print(participant.get("responses", {}).get("0", {}).get('final_version'), "\n\n\n")

test_ = participant.get("responses", {}).get("0", {}).get('final_version')
print(luar_embedding(test_, len(test_) + 1))
print(luar_embedding(test_, len(test_) + 1).shape)

['user_info', 'responses', 'pre_survey', 'mid_survey_1', 'mid_survey_2', 'post_survey'] 

f5d09877-f461-4169-a291-79c17218071d
['0', '1', '2', '3', '4', '5']
['start_time', 'scenario', 'model_generation_shown', 'details', 'model_generation', 'submit_details_time', 'final_version', 'edits', 'submit_final_text_time', 'likert'] 


Hello Paul! Hope you are doing great. I feel privileged to express my gratitude for what you did when i lost my mum. I can't imagine the financial support you offered me. Not forgeting the moral support you accorded me. By then my spirits were real low but you uplifted them. I can't give enough through my writing but you made me feel a very important person when i had just lost all the hopes. Thank you so much Paul for standing with me during that trying moment. I know i can accumulate enough to repay you for the hospitality but i felt honoured in your presence. You made me gain confident that had dwindled away. I really felt proud by your presence during that t

In [27]:
print(train_files[:10])

[WindowsPath('d:/Academics Semester 5/DS501 - Machine Learning/Project/PersonalizationOfLLMs/personal_style_postedit/logs/b417dba7-2b67-4496-b985-fc67c15aa748.json'), WindowsPath('d:/Academics Semester 5/DS501 - Machine Learning/Project/PersonalizationOfLLMs/personal_style_postedit/logs/32a9cad7-5b1c-480f-832c-1cedb782d226.json'), WindowsPath('d:/Academics Semester 5/DS501 - Machine Learning/Project/PersonalizationOfLLMs/personal_style_postedit/logs/99d7de62-94ed-4cc5-99bf-dbb5974df705.json'), WindowsPath('d:/Academics Semester 5/DS501 - Machine Learning/Project/PersonalizationOfLLMs/personal_style_postedit/logs/a03e9b46-d5c1-401e-ba35-3746d2880c76.json'), WindowsPath('d:/Academics Semester 5/DS501 - Machine Learning/Project/PersonalizationOfLLMs/personal_style_postedit/logs/6998773e-d1b4-4b6d-ac29-f75d132ecb89.json'), WindowsPath('d:/Academics Semester 5/DS501 - Machine Learning/Project/PersonalizationOfLLMs/personal_style_postedit/logs/21d20b0e-b89e-42a8-b0bb-cb8b2d2e35df.json'), Win

### 2.2 - Participant Level Structure

Given a json file path = train_files[i] or test_files[j] or json_files[k]

1. We would do -> with open(path, "r", encoding = "utf-8") as f: data = json.load(f)
2. Go to data['responses']['i'] for each i = 0, 1, 2, 3, 4, 5
3. Find data['responses']['i']['model_generation'] and ['final_version']
4. Keep ['model_generation'] texts as a list **ai_texts**
5. Divide ['final_version'] by ['responses']['i']['model_generation_shown']
6. Store them in lists **human_texts** and **treatment_texts**.

7. Concatenate AI texts into string **ai_combined = "\n\n".join(ai_texts)**
8. Concatenate human texts into string **human_combined = "\n\n".join(human_texts)**

9. Find embeddings of them. **ai_embedding and human_embedding**.
10. List of human edited texts embeddings as **treatment_embeddings**.

In [28]:
def progress_to_human(path, print_ans = False):
    
    with open(path, "r", encoding = "utf-8") as f:
        data = json.load(f)
    
    ai_texts = []
    human_texts = []
    treatment_texts = []
    
    for i in range(6):
        ai_texts.append(data["responses"][str(i)]["model_generation"])
        
        if data["responses"][str(i)]["model_generation_shown"] == 0:
            human_texts.append(data["responses"][str(i)]["final_version"])
        else:
            treatment_texts.append(data["responses"][str(i)]["final_version"])
    
    
    ai_combined = "\n\n".join(ai_texts)
    human_combined = "\n\n".join(human_texts)
    
    ai_embedding = luar_embedding(ai_combined, 512)
    human_embedding = luar_embedding(human_combined, 512)
    
    treatment_embeddings = []
    for txt in treatment_texts:
        treatment_embeddings.append(luar_embedding(txt, 512))
    
    
    '''Calculation 1 - Initial AI - Human similarity relationship'''
    similarity = cosine_sim(ai_embedding, human_embedding)
    treatment_sim = []
    for e in treatment_embeddings:
        treatment_sim.append(cosine_sim(e, human_embedding))
        
    
    changes = [s - similarity for s in treatment_sim]

    mean_change = sum(changes) / len(changes)

    if print_ans:
        print("Similarity between AI and Pure Human: ", similarity)
        print("Similarities between final edited draft and pure human: ")
        print(treatment_sim)
        print("Changes:", changes)
        print("Mean change:", mean_change)
    
    return (similarity.item(), 
        [x.item() for x in treatment_sim], 
        [x.item() for x in changes], 
        mean_change.item()
    )

In [29]:
def cosine_sim(a, b):
    dot_prd = torch.dot(a, b)
    den = torch.norm(a)*torch.norm(b)
    ans = dot_prd/den
    return ans

In [30]:
similarity, treatment_sim, changes, mean_change = progress_to_human(path)

print("AI-Human similarity:", similarity)
print("Treatment-Human similarities:", treatment_sim)
print("Changes:", changes)
print("Mean change:", mean_change)

AI-Human similarity: 0.5186756253242493
Treatment-Human similarities: [0.5923218727111816, 0.4786105751991272, 0.5149737000465393, 0.4246005117893219]
Changes: [0.07364624738693237, -0.04006505012512207, -0.003701925277709961, -0.09407511353492737]
Mean change: -0.016048960387706757


In [32]:
df = pd.DataFrame(columns = ["participant_id", "ai-human", "treatment-human", "changes", "mean_change"])

for prtc in train_files:
    similarity, treatment_sim, changes, mean_change = progress_to_human(prtc)
    df.loc[len(df)] = [prtc, similarity, treatment_sim, changes, mean_change]

In [33]:
df.to_csv(r"D:\Academics Semester 5\DS501 - Machine Learning\Project\PersonalizationOfLLMs\6j10a_df.csv")

In [34]:
df["mean_change"].describe()

count    57.000000
mean     -0.026449
std       0.072013
min      -0.209522
25%      -0.061263
50%      -0.022628
75%       0.008507
max       0.204432
Name: mean_change, dtype: float64

In [35]:
print("Mean:", df["mean_change"].mean())
print("Median:", df["mean_change"].median())
print("Positive:", (df["mean_change"] > 0).sum())
print("Negative:", (df["mean_change"] < 0).sum())

Mean: -0.026449080491275118
Median: -0.022628381848335266
Positive: 17
Negative: 40


Our assessment is showing that post human editing is moving away from the human reference on average.


In [36]:
df[["change_1", "change_2", "change_3", "change_4"]] = pd.DataFrame(
    df["changes"].tolist(),
    index=df.index
)

In [37]:
df[["change_1", "change_2", "change_3", "change_4"]].describe()

,change_1,change_2,change_3,change_4
count,57.000000,57.000000,57.000000,57.000000
mean,-0.031707,-0.031032,-0.021110,-0.021947
std,0.093760,0.099132,0.090280,0.093415
min,-0.208190,-0.273461,-0.252247,-0.211974
25%,-0.073847,-0.084437,-0.069600,-0.077491
50%,-0.034342,-0.037076,-0.011225,-0.009671
75%,0.020685,0.013568,0.037627,0.035716
max,0.178268,0.297796,0.271478,0.226092


#### Why our result can be negative?

